### PR TITLE
CPM-1045: On product save, update/fill the new completeness table

### PR DESCRIPTION
**Pull request by antoinearrive**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Completeness/DoubleStorageCompletenesses.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        $this->saveAll([$completenesses]);
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public function saveAll(array $productCompletenessCollections): void
+    {
+        $newTableExists = $this->newTableExists();
+        if ($newTableExists) {
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Completeness/DoubleStorageCompletenesses.php:40]:
```suggestion
        if ($this->newTableExists()) {
```

File: src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Completeness/DoubleStorageCompletenesses.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            $this->saveProductCompletenesses->saveAll($productCompletenessCollections);
+        }
+
+        $this->legacySaveProductCompletenesses->saveAll($productCompletenessCollections);
+
+        if (!$newTableExists) {
+            throw new \Exception(sprintf('The table "%s" is not present on database', self::TABLE_NAME));
+        }
```
Comment (mmetayer) [inline src/Akeneo/Pim/Enrichment/Bundle/Storage/Sql/Completeness/DoubleStorageCompletenesses.php:48]: no, we don't want to throw an exception here, the table will NOT exist during a few minutes (until the migration is run)

File: tests/back/Pim/Enrichment/Specification/Bundle/Storage/Sql/Completeness/DoubleStorageCompletenessesSpec.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    ): void
+    {
+        $connection->executeQuery(Argument::type('string'), Argument::type('array'))->willReturn($result);
+        $result->rowCount()->willReturn(0);
+
+        $completeness = new ProductCompletenessWithMissingAttributeCodesCollection('25d9c9f8-bfb0-4406-88d9-76af05ee3cde', [
+            new ProductCompletenessWithMissingAttributeCodes('ecommerce', 'en_US', 5, ['name', 'description']),
+        ]);
+
+        $this->shouldThrow(\Exception::class)->during('save', [$completeness]);
```
Comment (mmetayer) [inline tests/back/Pim/Enrichment/Specification/Bundle/Storage/Sql/Completeness/DoubleStorageCompletenessesSpec.php:74]: nope

**Review by mmetayer — APPROVED** [review]

File: tests/back/Pim/Enrichment/Integration/Storage/Sql/Completeness/SqlSaveCompletenessesIntegration.php [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                        'missing' => 0,
+                        'required' => 5,
+                    ]
+                ]
+            ],
+            $dbCompletenesses,
+        );
+    }
+
+    public function test_that_it_saves_completenesses()
```
Comment (Beninho) [inline tests/back/Pim/Enrichment/Integration/Storage/Sql/Completeness/SqlSaveCompletenessesIntegration.php:40]:
```suggestion
    public function test_that_it_saves_completenesses(): void
```

File: tests/back/Pim/Enrichment/Specification/Bundle/Storage/Sql/Completeness/DoubleStorageCompletenessesSpec.php [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        Connection $connection,
+    ) {
+        $this->beConstructedWith(
+            $legacySaveProductCompletenesses,
+            $saveProductCompletenesses,
+            $connection,
+        );
+    }
+
+    function it_is_initializable()
```
Comment (Beninho) [inline tests/back/Pim/Enrichment/Specification/Bundle/Storage/Sql/Completeness/DoubleStorageCompletenessesSpec.php:34]:
```suggestion
    function it_is_initializable(): void
```

**Review by Beninho — APPROVED** [review]

File: tests/back/Pim/Enrichment/Integration/Storage/Sql/Completeness/SqlSaveCompletenessesIntegration.php (hunk 4 above)
Comment (mmetayer) [reply tests/back/Pim/Enrichment/Integration/Storage/Sql/Completeness/SqlSaveCompletenessesIntegration.php:40]: we will fix this in #19714

**Pull request merged**